src/text-buffer.coffee

```diff
@@ -843,19 +842,15 @@ class TextBuffer
 
   # Public: Undo the last operation. If a transaction is in progress, aborts it.
   undo: ->
-    markerSnapshot = _.extend(@markerStore.createSnapshot(), @lastMarkerSnapshot)
-    if pop = @history.popUndoStack(markerSnapshot)
+    if pop = @history.popUndoStack(@markerStore.createSnapshot())
```

Comment:
Maybe not now, but I wonder if you could save any time by creating the snapshot with knowledge of the existing undo/redo snapshot you're going to assign into via `_.defaults`. You would only need to do work for markers that weren't already present.